enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,210 @@
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will have a dependency on the `greenboot` RPM
+and it will include a health check script to validate that all the required
+MicroShift services are up and running. The script will be installed into the
+`/etc/greenboot/check/required.d` directory and it will not proceed with the
+validation in case the MicroShift service was not enabled or the system was not
+booted via `ostree`.
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+The health check script will implement the following validations for the core
+MicroShift namespaces:
+- openshift-ovn-kubernetes
+- openshift-service-ca
+- openshift-ingress
+- openshift-dns
+- openshift-storage
+
+> When pluggable CNI and CSI functionality is implemented, the `openshift-ovn-kubernetes`
+> and `openshift-storage` namespace name selection need to be conditional on
+> the loaded CNI and CSI plugins.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+All the script output is logged at `/var/log/microshift-greenboot.log` file.
+
+|Validation                                        |Pass  |Fail  |
+|--------------------------------------------------|------|------|
+|Check the system was booted via ostree            |Next  |exit 0|
+|Check microshift.service is enabled               |Next  |exit 0|
+|Wait for microshift.service to be active (!failed)|Next  |exit 1|
+|Wait for any pods to start                        |Next  |exit 1|
+|For each core namespace, wait for pods to be ready|Next  |exit 1|
+|For each core namespace, check pods not restarting|exit 0|exit 1|
+
+The wait period in each validation starts from 5 minutes base time and it is
+incremented by the base wait period after each boot in the verification loop.
+It should be possible to override the base time wait period setting with the
+`MICROSHIFT_WAIT_TIMEOUT_SEC` environment variable.
+
+### Risks and Mitigations
+
+Edge Devices often have relatively slow networks or occasional loss of network
+connectivity. The risk of declaring a `false positive` on the failed boot may
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, system administrators should set the base
+wait period setting to a higher value to start with.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, system administrators should reinstall the upgrade when network
```

Comment:
How does system admins know the upgrade failure is caused by slow network? Does the greenbook report failure with a reason?

Reply:
That's a good point, we do not report reasons at present. Any suggestions on how to improve this?

Reply:
When greenboot scripts are run as part of the boot procedure, all the output is logged to the system log. I removed explicit log file from the proposal and changed the text for users to look at the system log.